File: lib/models/gamification/nudge_label.dart
class NudgeLabel {
  final String key;
  final Map<String, dynamic>? params;

  /// Server copy fallback when localization bundle has no entry (see `default_text` in API).
  final String? defaultText;

  const NudgeLabel({required this.key, this.params, this.defaultText});

  /// [raw] is either `{ "key", "params" }` or a legacy plain string (deprecated).
  factory NudgeLabel.fromDynamic(dynamic raw, {String? legacyTitle}) {
    if (raw == null) {
      final t = legacyTitle;
      if (t != null && t.isNotEmpty) {
        return NudgeLabel(key: 'legacy_literal', params: {'text': t});
      }
      return const NudgeLabel(key: '', params: null);
    }
    if (raw is Map<String, dynamic>) {
      final k = raw['key'];
      return NudgeLabel(
        key: k is String ? k : '',
        params: _normalizeParams(_castParams(raw['params'])),
        defaultText: raw['default_text'] as String? ?? raw['defaultText'] as String?,
      );
    }
    if (raw is Map) {
      final k = raw['key'];
      return NudgeLabel(
        key: k is String ? k : '',
        params: _normalizeParams(_castParams(raw['params'])),
        defaultText: raw['default_text'] as String? ?? raw['defaultText'] as String?,
      );
    }
    if (raw is String) {
      return NudgeLabel(key: 'legacy_literal', params: {'text': raw});
    }
    return const NudgeLabel(key: '', params: null);
  }

  static Map<String, dynamic>? _castParams(dynamic p) {
    if (p == null) return null;
    if (p is Map<String, dynamic>) return p;
    if (p is Map) {
      return p.map((k, v) => MapEntry('$k', v));
    }
    return null;
  }

  /// Duplicate snake_case keys as camelCase so `{{redCards}}` templates work with Python-style JSON.
  static Map<String, dynamic>? _normalizeParams(Map<String, dynamic>? p) {
    if (p == null) return null;
    final out = Map<String, dynamic>.from(p);
    if (out.containsKey('red_cards') && !out.containsKey('redCards')) {
      out['redCards'] = out['red_cards'];
    }
    if (out.containsKey('gold_coins') && !out.containsKey('goldCoins')) {
      out['goldCoins'] = out['gold_coins'];
    }
    if (out.containsKey('deadline_time') && !out.containsKey('deadlineTime')) {
      out['deadlineTime'] = out['deadline_time'];
    }
    return out;
  }

  bool get isValid => key.isNotEmpty;
}

File: shared/src/commonMain/kotlin/com/snabbit/runner/shared/features/gamification/domain/model/NudgeLabel.kt
package com.snabbit.runner.shared.features.gamification.domain.model

/**
 * A localisation token — the KMP port of Dart `lib/models/gamification/nudge_label.dart`.
 *
 * A label is either a server-driven `{ key, params, default_text }` triple or a
 * legacy plain string (wrapped as `key == `[LEGACY_LITERAL]`, params = {"text": …}`).
 * [params] values are stringified at parse time so `{{placeholder}}` templating
 * is a pure string substitution. The parser duplicates `red_cards` / `gold_coins`
 * / `deadline_time` into their camelCase forms so Python-style snake_case payloads
 * still satisfy `{{redCards}}`-style templates.
 *
 * Resolution to display text (server `default_text` → client English fallback)
 * lives in the presentation layer (`ResolveNudgeLabel`), matching Dart.
 */
data class NudgeLabel(
    val key: String,
    val params: Map<String, String> = emptyMap(),
    val defaultText: String? = null,
) {
    /** Mirrors Dart `NudgeLabel.isValid`. */
    val isValid: Boolean get() = key.isNotEmpty()

    companion object {
        const val LEGACY_LITERAL = "legacy_literal"

        /** Empty / invalid label — the parse fallback for a null or malformed value. */
        val EMPTY = NudgeLabel(key = "")

        /** Wrap a legacy plain-string title as a literal label. */
        fun literal(text: String): NudgeLabel =
            NudgeLabel(key = LEGACY_LITERAL, params = mapOf("text" to text))
    }
}
